### ISSUE TITLE
Keycloak provider v7.2.1 token validation request failed

### ISSUE THREAD

**Issue by tatianaschaefer0815** [oauth2-proxy/oauth2-proxy#1599]
After the update from v7.2.0 to v7.2.1 the keycloak has an issue with token validation

## Current Behavior
The container produces these logs:

```
[2022/03/23 09:43:58] [internal_util.go:69] 400 GET https://<keycloak-url>/auth/realms/<realm>/protocol/openid-connect/userinfo?access_token=<some-token>... {"error":"invalid_request","error_description":"Token not provided"}
[2022/03/23 09:43:58] [internal_util.go:74] token validation request failed: status 400 - {"error":"invalid_request","error_description":"Token not provided"}
127.0.0.1:53088 - b7491af6-3cb5-4f67-b844-0291d651133b - james.holden@expanse.de [2022/03/23 09:43:59] [AuthFailure] Session validation failed: Session{email:james.holden@expanse.de user: PreferredUsername: token:true created:2022-03-23 09:43:48.629693961 +0100 CET m=+96.088076113 expires:2022-03-23 11:43:48.629693961 +0100 CET m=+7296.088076113}
```

configuration:

```
--https-address=0.0.0.0:8443
--upstream=http://<some-host>:9090
--tls-cert-file=/tmp/cert/cert.pem
--tls-key-file=/tmp/cert/key.pem
--request-logging=true
--auth-logging=true
--skip-provider-button
--cookie-expire=2h
--cookie-name=<cookie-name>
--skip-jwt-bearer-tokens
--extra-jwt-issuers="https://<keycloak>/auth/realms/<realm>=realm-management"
--provider=keycloak
--login-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/auth
--redeem-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/token
--validate-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/userinfo
--scope="openid profile groups email"
--email-domain="expanse.de"
--client-secret=<secret>
--client-id=<client>
--cookie-secret=<cookie>
```

## Expected Behavior
We expect that the Keyclaok provider is able to validate the token

## Possible Solution

Please add Session validation with header to keycloak.go
```
//ValidateSession validates the AccessToken
func (p *KeycloakProvider) ValidateSession(ctx context.Context, s *sessions.SessionState) bool {
	return validateToken(ctx, p, s.AccessToken, makeOIDCHeader(s.AccessToken))
}
```

**Comment by DmytroKozlovskyi** [comment]
keycloak-oidc also doesn't work for us.

**Comment by JoelSpeed** [comment]
> Upd: Actually keycloak-oidc works after remove scope = "email"

So did you find a solution? Can we close this issue?

**Comment by Beanow** [comment]
I'm seeing the same issue, but didn't find a scope combination that works. But reverting to docker tags `v7.1.3` or `v7.2.0` does work. Note the 400 error internally also happens with the older versions. But the session validation doesn't fail, unlike `v7.2.1`

This is from `v7.2.0`:
```
[2022/04/19 18:23:37] [internal_util.go:69] 400 GET https://<keycload>/auth/realms/<realm>/protocol/openid-connect/userinfo?access_token=eyJ... {"error":"invalid_request","error_description":"Token not provided"}
<ip> - <guid> - <email> [2022/04/19 18:23:37] [AuthSuccess] Authenticated via OAuth2: Session{email:<email> user: PreferredUsername: token:true created:2022-04-19 18:23:37.909707901 +0000 UTC m=+46.231369998 expires:2022-04-20 02:23:37.909707901 +0000 UTC m=+28846.231369998}
```

Meanwhile `v7.2.1`:
```
[2022/04/19 18:33:19] [internal_util.go:69] 400 GET https://<keycload>/auth/realms/<realm>/protocol/openid-connect/userinfo?access_token=eyJ... {"error":"invalid_request","error_description":"Token not provided"}
[2022/04/19 18:33:19] [internal_util.go:74] token validation request failed: status 400 - {"error":"invalid_request","error_description":"Token not provided"}
<ip> - <guid> - <email> [2022/04/19 18:33:19] [AuthFailure] Session validation failed: Session{email:<email> user: PreferredUsername: token:true created:2022-04-19 18:33:19.172137837 +0000 UTC m=+158.353466332 expires:2022-04-20 02:33:19.172137837 +0000 UTC m=+28958.353466332}
```

Relevant config:

```yml
version: '3.8'
services:
  oidc-gate:
    image: quay.io/oauth2-proxy/oauth2-proxy:v7.2.1
    environment:
      OAUTH2_PROXY_COOKIE_SECRET: xxxxx
      OAUTH2_PROXY_CLIENT_SECRET: xxxxx
    command:
      - --force-https=false
      - --flush-interval=1ms
      - --reverse-proxy=true
      - --http-address=0.0.0.0:3000
      - --upstream=http://caddy:80
      - --cookie-secure=true
      - --cookie-expire=8h
      - --cookie-domain=xxxx
      - --provider=keycloak
      - --provider-display-name=xxxx
      - --client-id=xxxx
      - --login-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/auth
      - --redeem-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/token
      - --validate-url=https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/userinfo
      - --scope=email
      - --user-id-claim=email
      - --prefer-email-to-user=true
      - --email-domain=*
      - --skip-provider-button=true
```

**Comment by Beanow** [comment]
I think this may be similar to the issue to https://github.com/oauth2-proxy/oauth2-proxy/issues/1505
And that was made apparent by https://github.com/oauth2-proxy/oauth2-proxy/pull/1433

Looks like this "400 Invalid Request - Token not provided" error was present already.
But since `v7.2.1` that results in failed authentication instead of being ignored.

**Comment by JoelSpeed** [comment]
Ok so it looks like the change here is because we are trying to align the way that we extract information from user info. In some cases, an error caused an auth failure, in others, the auth was allowed through. We aligned on the closed route for now as that is more secure. Looks to me like we need to update the keycloak provider to pass the correct authentication to the userinfo URL, I'm guessing it wants the access token to be passed but we are not doing that right now?

**Comment by tatianaschaefer0815** [comment]
the bug seems to have been fixed with version 7.3.0
